contracts/protocore/SelfProtocore.sol

```diff
@@ -69,5 +69,42 @@ contract SelfProtocore is MasterCopyNonUpgradable, GenesisSelfProtocore, Validat
             genesisAuxiliaryTargetBlockHash,
             genesisAuxiliaryTargetBlockNumber
         );
+        ValidatorSet.setupValidatorSet();
+    }
+
+
+    /* External Functions */
+
+    /**
+     * @notice Insert or remove validator. It inserts validator if not already
+     *         present and reputation is greater than 0. It removes validator
+     *         if it is present and reputation is 0.
+     *
+     * @dev Function requires:
+     *      - Caller should be Coconsensus contract.
+     *      - Validator must enter at height equal to current open kernel height.
+     *
+     * @param _validator Validator address to upsert.
+     * @param _height Validator start or end height to be updated.
+     * @param _reputation Validator's reputation value.
+     */
+    function upsertValidator(
+        address _validator,
+        uint256 _height,
+        uint256 _reputation
+    )
+        external
+        onlyCoconsensus
+    {
+        if(validators[_validator] != address(0)) {
```

Comment:
doesn't this respect the interface better (aka `validators[]` should probably just be private and an `external view` function iterates over (some length) to return an array

```
if (isInValidatorSet(_validator, _height)) {...}
```

Reply:
we can add an `assert(_height == openKernelHeight.add(1))`

Reply:
Thank you.
Added check `isInValidatorSet` and suggested `assert`. 
Yes, I liked your suggestion to make validators private and have an iterator function. Shall I create the ticket to address it ?

Reply:
yes, please make the ticket 💪

Im not resolving this because I think the `assert` should be at the beginning of the function

Reply:
Noted the point to make ticket.